### DISCUSSION TITLE
possible bug

### DISCUSSION THREAD

**Discussion by PaltryProgrammer** [boostorg/safe_numerics#137] [Experiences with safe_numerics]
greetings kind regards
attached please find PNGs of source and output of `auto sum = uimax + imin;` result was exception as shown . the correct result id est 9,223,372,036,854,775,807 i believe fits both `__int64` and `unsigned __int64` .
attached also  is source for purpose of reproduction .
 thank you kindly
[JunkCode.zip](https://github.com/boostorg/safe_numerics/files/12708754/JunkCode.zip)

![image](https://github.com/boostorg/safe_numerics/assets/48535920/cd094c48-3a61-4850-965a-f0a28b8f03a0)
![image](https://github.com/boostorg/safe_numerics/assets/48535920/f695048a-dc5f-439e-adc1-db00ec887e6b)

**Comment by robertramey** [comment]
C++ promotion rules state that the procedure for arithmetic operations is:
a) convert operands to the same type using rules specified in the C++ standard.  For addition of signed and unsigned types, the signed type is converted to unsigned (I know).  In your case this results in value that cannot be represented as an unsigned type.  When using safe integer types, this will be detected as an error.
b) perform the operation - code never arrives here.

**Reply by PaltryProgrammer** [reply]
my apologies
i forgot the result type can not be determined at run time .

**Comment by PaltryProgrammer** [comment]
thank you kindly for reply
please permit me to re-state am utilizing `automatic` promotion rule which i assume guarantees result given a type exists which can hold it . as stated for the given value i believe both `intmax_t` and `uintmax_t` can .

**Reply by robertramey** [reply]
automatic promotion rule which i assume guarantees result given a type exists.

Nope. It calculates the range of the result.  Then it selects a type which is guaranteed to hold that range so no runtime checking is necessary.  BUT if no such type exists, it selects the largest type available AND checks the result at runtime.

If you want to guarantee no overflow at runtime, use automatic AND set the exception policy to "trap".  This will create a runtime error any time a value COULD overflow.  So no runtime checking is required.
